controllers/shooter.py

```diff
@@ -75,54 +69,54 @@ def searching(self) -> None:
         if self.vision.get_vision_data()[2] is not None:
             # means no data is available
             # print(f"searching -> tracking {self.vision.get_vision_data()}")
-            # self.next_state("tracking")
-            self.state = self.tracking
+            self.next_state("tracking")
 
-    # @state
-    def tracking(self) -> None:
+    @state
+    def tracking(self, initial_call) -> None:
         """
         Aiming towards a vision target and spining up flywheels
         """
+        if initial_call:
+            self.shooter.stop_motors()
         dist, delta_angle, timestamp = self.vision.get_vision_data()
         # collect data only once per loop
         if timestamp is None:
-            # self.next_state("searching")
+            self.next_state("searching")
             # print(f"tracking -> searching {self.vision.get_vision_data()}")
-            self.state = self.searching
         else:
             if abs(delta_angle) > self.find_allowable_angle(dist):
                 # print(f"Telling turret to slew by {delta_angle}")
                 self.turret.slew(delta_angle)
             if self.ready_to_spin():
-                # self.next_state("firing")
-                # print(f"tracking -> spining_up {self.vision.get_vision_data()}")
                 self.distance = dist
-                self.state = self.spining_up
+                self.next_state("spining_up")
+                # print(f"tracking -> spining_up {self.vision.get_vision_data()}")
 
-    def spining_up(self) -> None:
-        if self.initial_call:
+    @state
+    def spining_up(self, initial_call) -> None:
+        if initial_call:
             self.shooter.set_range(self.distance)
-            self.initial_call = False
-        if self.ready_to_fire() and self.input_command:
-            self.distance = None
-            self.initial_call = True
-            # print(f"spining_up -> firing {self.vision.get_vision_data()}")
-            self.state = self.firing
-
-    # @state
+        if self.turret.is_ready():
```

Comment:
We get to this state only if ready_to_spin returns true, and one of its conditions is that the turret is ready, so this check seems unnecessary. Is there a situation in which the turret could have been ready, we starting spinning up the shooter, then the turret stopped being ready? My understanding is that we now want to slew the turret only once after we've stopped moving, then spin up and fire. Spinning can overlap slewing once we have set the slewing target, but once we are spinning we can't get any more vision data, so we can't slew until we've stopped spinning. So I don't think we need this check. Or am I missing something?

Reply:
Hmm. On reflection, ready_to_spin should not depend on the turret being ready, just on the turret having started a slew to firing position. That way we can overlap spinning up the shooter motors with slewing the turret.

Reply:
This might require more turret API, to determine if we are slewing, and we might need to distinguish between slewing as part of a scan and slewing to a target. Think through this state machine and let me know what you need. Oh, and while you are at it, we can't implement scan in the turret because the turret knows nothing about vision data. Scanning for a target has to be part of the shooter controller state machine and just use turret slewing commands.